alerts/machine-config-operator/MachineConfigControllerDrainError.md

```diff
@@ -0,0 +1,114 @@
+```
+
+Finally if 1 hour has passed and the drain is still failing the following logs will be fired and the node will be marked degraded.
+
+```console
+1 drain_controller.go:352] node xxxxxxx-xxxxxx-xxxxxx-xxxxx: drain exceeded timeout: 1h0m0s. Will continue to retry.
+```
+
+```console
+1 status.go:126] Degraded Machine: xxxxxxx-xxxxxx-xxxxxx-xxxxx: and Degraded Reason: failed to drain node: xxxxxxx-xxxxxx-xxxxxx-xxxxx after 1 hour. Please see machine-config-controller logs for more information
+```
+
+The machine-controller logs will explicitly tell you what pods are failing to drain succesfully. Looking at the pods named in the logs is where you should start. 
+
+Common reasons a node cannot drain a pod successfully are:
+
+- The pod has a PodDisruptionBudget (PBD) set preventing the controller from deleting the pod.
+- The pod has storage attached and the kubelet is unable to unmount the storage succesfully.
+- The pod deletion may be prevented by a webhook if it's configured to target UPDATE operations or the webhook is failing to be called by the kube-apiserver.
+- The pod deletion could be stuck because the finalizers set in the pod are preventing it from terminating succesfully.
+
+## Mitigation 
+
+- If a PDB is causing the failure you can temporarily patch it to set the `minAvailable` to 0 so it can scale the pod down successfully. Then patch it to the old value after the upgrade completes.
+
+```console
+oc patch pdb $PDB -n $NS --type=merge -p '{"spec":{"minAvailable":0}}'
+```
+- If a webhook is preventing the pod deletion you can check the `openshift-kube-apiserver` logs to see exactly what webhook is preventing deletion.
+
+```console
+Failed calling webhook, failing open vault.hashicorp.com: failed calling webhook "webhook_name": failed to call webhook: Post "https://hashi-vault-agent-injector-svc.vault-injector.svc:443/mutate?timeout=30s": context canceled2022-10-11T19:13:53.345381984Z E1011 19:13:53.345348      16 dispatcher.go:184] failed calling webhook "webhook_name": failed to call webhook: Post "https://name-injector-svc.vault-injector.svc:443/mutate?timeout=30s": context canceled
+```
+Then check if it's a `mutating` or `validating` webhook.
+```console
+$ oc get validatingwebhookconfiguration
+$ oc get mutatingwebhookconfiguration
+```
+Then backup and delete the webhook 
+```console
+$ oc get validatingwebhookconfiguration/<webhook_name> -o yaml > webhook.yaml
+$$ oc delete <webhook_type> <webhook_name>         
+```
+This should allow the drain to continue and if the webhook does not come back automatically you can recreate it via the backup.
+
+- If a pod cannot amount storage you either need to troubleshoot why it's failing such as if it's a NFS storage then it could be a network issue with the storage server. If you are okay with possible data loss you can force delete the pod and immediately remove resources from the API and bypass graceful deletion. 
```

Comment:
```suggestion
- If a pod cannot unmount storage, troubleshoot why it's failing. For example, if you are using NFS storage, the problem could be a network issue with the storage server. 

Otherwise, if you are comfortable with possible data loss, you can force delete the pod and immediately remove resources from the API and bypass graceful deletion by using the following command: 
```

Reply:
Agreed, Applied the change